### PR TITLE
Alignment

### PR DESCRIPTION
**Pull request by oli-obk**
Every allocation now wastes `alignment` bytes. This is so the `allocate` method always returns a pointer that is just aligned enough for the requirement, but not more. This way, if you allocate a `[u32, 2]` you can't cast a pointer to it to `*mut u64` and write to it (except on 32 bit linux)

### CODE REVIEW

File: src/interpreter/mod.rs [hunk 1]
```diff
@@ -176,19 +177,19 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
         use rustc_const_math::{ConstInt, ConstIsize, ConstUsize, ConstFloat};
         macro_rules! i2p {
             ($i:ident, $n:expr) => {{
-                let ptr = self.memory.allocate($n)?;
+                let ptr = self.memory.allocate($n, $n)?;
```
Comment (oli-obk) [inline src/interpreter/mod.rs]: This over-aligns `u64` on linux, and might underalign on some hypothetical system where the alignment of primitives is bigger than the size, but if that happens, the write will fail, so let's be lazy here :)

**Comment by oli-obk** [comment]
ping?

File: src/memory.rs [hunk 2]
```diff
@@ -50,6 +51,25 @@ impl Pointer {
             offset: 0,
         }
     }
+    pub fn is_aligned_to(&self, align: usize) -> bool {
```
Comment (solson) [inline src/memory.rs]: I'm wondering if this should take the `self.align` into account (i.e. the alignment of the base of the allocation). For example, if some pointer `p` was 2-aligned and at offset 0, then `p.is_aligned_to(4)` returns true, but it wouldn't necessarily be true in reality. A 2-aligned pointer *could* be 4-aligned, but it could also not be.

File: src/memory.rs [hunk 3]
```diff
@@ -133,69 +154,79 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
...
     }
 
-    pub fn allocate(&mut self, size: usize) -> EvalResult<'tcx, Pointer> {
+    pub fn allocate(&mut self, size: usize, align: usize) -> EvalResult<'tcx, Pointer> {
         if size == 0 {
             return Ok(Pointer::zst_ptr());
         }
+        // make sure we can offset the result pointer by the worst possible alignment
+        // this allows cheaply checking for alignment directly in the pointer
+        let least_aligned_size = size + align;
```
Comment (solson) [inline src/memory.rs]: I don't really understand this, but it seems fishy to me. The first thing I noticed is that a simple array of chars (say, size 10, align 1) will have an extra byte for no reason. Can we accomplish this without wasting any bytes?

File: src/memory.rs [hunk 4]
```diff
@@ -133,69 +154,79 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
...
+            undef_mask: UndefMask::new(least_aligned_size),
+            align: align,
         };
         let id = self.next_id;
         self.next_id.0 += 1;
         self.alloc_map.insert(id, alloc);
         Ok(Pointer {
             alloc_id: id,
-            offset: 0,
+            // offset by the alignment, so larger accesses will fail
```
Comment (solson) [inline src/memory.rs]: Maybe we should just cache the allocation's alignment in the pointer? Not sure...

**Comment by solson** [comment]
Sorry for the delay in reviewing this.

I don't like the approach of wasting bytes. I'm not sure about the best way to handle this, but I did have an interesting thought...

On a traditional machine, one could take a pointer to some 4-aligned `[u32; 3]` and then either `&p[0]` or `&p[1]` would be 8-aligned (you could check by casting the base pointer to `usize` and using `base % 8 == 0`) so you could take one of those two pointers, cast to `*mut u64`, and write successfully.

But in Miri, it doesn't necessarily need to be observable what an allocation's absolute alignment is. In the 4-aligned `[u32; 3]` example, if you take this PR's approach, then the base pointer will never be 8-aligned but the `&p[1]` pointer will _always_ be. My thought is that we could find a way to reject all attempts to use any part of the 4-aligned allocation as 8-aligned.

Another consequence of my approach would be that writing _anything_ to _any_ part of a `Vec`'s data array would fail if it required alignment greater than 1. Under this PR's behaviour there would be a point where any given alignment could write to the data buffer (e.g. an 8-aligned value could always be written at offset 7).

As a final note, I would allow going the other direction. That is, a 4-aligned allocation would permit 2-aligned accesses on the multiple-of-2 offsets.

cc @eddyb Do you have any problems with my suggestion?

File: src/memory.rs (hunk 2 above)
Comment (solson) [inline src/memory.rs]: I think you can disregard this and just respond to my long comment on the PR. I didn't really grasp what was going on at first.

**Comment by oli-obk** [comment]
> I don't like the approach of wasting bytes

neither do I. I'm not sure how common `ptr % align == 0` checks are nowadays. I just remember seeing them in C-code all over the place.

I'll change it to the more complete checks, we can discuss with ppl when they hit the alignment issue

> As a final note, I would allow going the other direction. That is, a 4-aligned allocation would permit 2-aligned accesses on the multiple-of-2 offsets.

Oh we definitely need to do that, otherwise structs containing 4 byte and 2 byte aligned types wouldn't be usable.

**Comment by eddyb** [comment]
@oli-obk `(ptr as usize) % align == 0` in code evaluated by miri? That wouldn't possibly work with non-constant pointers.

File: src/memory.rs [hunk 5]
```diff
@@ -161,16 +165,17 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
 
     // TODO(solson): Track which allocations were returned from __rust_allocate and report an error
     // when reallocating/deallocating any others.
-    pub fn reallocate(&mut self, ptr: Pointer, new_size: usize) -> EvalResult<'tcx, Pointer> {
+    pub fn reallocate(&mut self, ptr: Pointer, new_size: usize, align: usize) -> EvalResult<'tcx, Pointer> {
+        // TODO(solson): Report error about non-__rust_allocate'd pointer.
```
Comment (solson) [inline src/memory.rs]: This comment is duplicated 2 lines down.

File: src/memory.rs [hunk 6]
```diff
@@ -179,15 +184,19 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
             alloc.bytes.extend(iter::repeat(0).take(amount));
             alloc.undef_mask.grow(amount, false);
         } else if size > new_size {
+            // it's possible to cause miri to use arbitrary amounts of memory that aren't detectable
+            // through the memory_usage value, by allocating a lot and reallocating to zero
             self.memory_usage -= size - new_size;
             self.clear_relocations(ptr.offset(new_size as isize), size - new_size)?;
             let alloc = self.get_mut(ptr.alloc_id)?;
             alloc.bytes.truncate(new_size);
-            alloc.bytes.shrink_to_fit();
```
Comment (solson) [inline src/memory.rs]: Why was this deleted?

**Comment by solson** [comment]
I'm ready to merge after my latest nits are addressed.

**Comment by solson** [comment]
This finishes the last item in https://github.com/solson/miri/issues/29.
